File: sqlsm/profiles.py
# -*- coding: utf-8 -*-
import json
import os
import time
import uuid
from typing import Any, Dict, List, Optional

from sqlsm.crypto_util import decrypt_secret, encrypt_secret

STORE_DIR = os.path.join(os.path.expanduser("~"), ".sqlsm")
STORE_PATH = os.path.join(STORE_DIR, "connections.json")


def _now():
    return time.strftime("%Y-%m-%d %H:%M:%S")


def _load():
    # type: () -> List[Dict[str, Any]]
    if not os.path.isfile(STORE_PATH):
        return []
    try:
        with open(STORE_PATH, "r", encoding="utf-8") as handle:
            data = json.load(handle)
    except Exception:
        return []
    items = data.get("profiles") if isinstance(data, dict) else data
    if not isinstance(items, list):
        return []
    return [item for item in items if isinstance(item, dict)]


def _save(items):
    # type: (List[Dict[str, Any]]) -> None
    if not os.path.isdir(STORE_DIR):
        os.makedirs(STORE_DIR)
    with open(STORE_PATH, "w", encoding="utf-8") as handle:
        json.dump({"profiles": items}, handle, ensure_ascii=False, indent=2)
    try:
        os.chmod(STORE_PATH, 0o600)
    except Exception:
        pass


def _key(item):
    return (
        str(item.get("server") or "").strip().lower(),
        str(item.get("port") or 1433),
        str(item.get("instance") or "").strip().lower(),
        str(item.get("auth") or "sql"),
        str(item.get("username") or "").strip().lower(),
        str(item.get("database") or "master").strip().lower(),
    )


def read_password(item):
    # type: (Dict[str, Any]) -> str
    raw = item.get("password") or ""
    if not raw:
        return ""
    return decrypt_secret(str(raw))


def public_profile(item):
    # type: (Dict[str, Any]) -> Dict[str, Any]
    auth = item.get("auth") or "sql"
    label = "%s · %s" % (item.get("server") or "server", item.get("database") or "master")
    if item.get("instance"):
        label = "%s\\%s · %s" % (item.get("server"), item.get("instance"), item.get("database") or "master")
    if auth == "windows":
        label += " · Windows"
    elif item.get("username"):
        label += " · " + item.get("username")
    return {
        "id": item.get("id"),
        "label": label,
        "server": item.get("server") or "",
        "port": item.get("port") or 1433,
        "instance": item.get("instance") or "",
        "auth": auth,
        "username": item.get("username") or "",
        "database": item.get("database") or "master",
        "encrypt": bool(item.get("encrypt")),
        "has_password": bool(item.get("password")),
        "remember_password": bool(item.get("remember_password")),
        "last_used": item.get("last_used") or "",
    }


def list_profiles():
    # type: () -> List[Dict[str, Any]]
    items = sorted(_load(), key=lambda item: item.get("last_used") or "", reverse=True)
    return [public_profile(item) for item in items]


def get_profile(profile_id):
    # type: (str) -> Optional[Dict[str, Any]]
    for item in _load():
        if item.get("id") == profile_id:
            return item
    return None


def upsert_profile(payload, remember_password=False):
    # type: (Dict[str, Any], bool) -> Dict[str, Any]
    items = _load()
    incoming = {
        "server": str(payload.get("server") or "").strip(),
        "port": int(payload.get("port") or 1433),
        "instance": str(payload.get("instance") or "").strip(),
        "auth": "windows" if payload.get("auth") == "windows" else "sql",
        "username": str(payload.get("username") or "").strip(),
        "database": str(payload.get("database") or "master").strip() or "master",
        "encrypt": bool(payload.get("encrypt")),
        "remember_password": bool(remember_password),
        "last_used": _now(),
    }
    password = str(payload.get("password") or "")
    match = None
    for item in items:
        if _key(item) == _key(incoming):
            match = item
            break
    if match is None:
        incoming["id"] = uuid.uuid4().hex[:12]
        if remember_password and password:
            incoming["password"] = encrypt_secret(password)
        items.insert(0, incoming)
        match = incoming
    else:
        match.update(incoming)
        if remember_password:
            if password:
                match["password"] = encrypt_secret(password)
        else:
            match.pop("password", None)
            match["remember_password"] = False
    _save(items)
    return public_profile(match)


def delete_profile(profile_id):
    # type: (str) -> bool
    items = _load()
    kept = [item for item in items if item.get("id") != profile_id]
    if len(kept) == len(items):
        return False
    _save(kept)
    return True

File: server/src/sql/profiles.ts
import { existsSync, mkdirSync, readFileSync, writeFileSync, chmodSync } from "node:fs";
import { join } from "node:path";
import { randomBytes } from "node:crypto";
import { secretKey, STORE_DIR } from "../config.js";

const STORE_PATH = join(STORE_DIR, "connections.json");
const PREFIX = "enc:v1:";

function encryptSecret(text: string): string {
  if (!text) return "";
  const key = secretKey();
  const data = Buffer.from(text, "utf8");
  const out = Buffer.alloc(data.length);
  for (let i = 0; i < data.length; i += 1) out[i] = data[i] ^ key[i % key.length];
  return PREFIX + out.toString("base64url");
}

export function decryptSecret(blob: string): string {
  if (!blob) return "";
  const text = String(blob);
  if (!text.startsWith(PREFIX)) return text;
  const key = secretKey();
  const data = Buffer.from(text.slice(PREFIX.length), "base64url");
  const out = Buffer.alloc(data.length);
  for (let i = 0; i < data.length; i += 1) out[i] = data[i] ^ key[i % key.length];
  return out.toString("utf8");
}

type Profile = Record<string, unknown>;

function load(): Profile[] {
  if (!existsSync(STORE_PATH)) return [];
  try {
    const data = JSON.parse(readFileSync(STORE_PATH, "utf8"));
    const items = Array.isArray(data) ? data : data.profiles;
    return Array.isArray(items) ? items.filter((item) => item && typeof item === "object") : [];
  } catch {
    return [];
  }
}

function save(items: Profile[]) {
  mkdirSync(STORE_DIR, { recursive: true });
  writeFileSync(STORE_PATH, JSON.stringify({ profiles: items }, null, 2));
  try {
    chmodSync(STORE_PATH, 0o600);
  } catch {
    /* ignore */
  }
}

function keyOf(item: Profile) {
  return [
    String(item.server || "").trim().toLowerCase(),
    String(item.port || 1433),
    String(item.instance || "").trim().toLowerCase(),
    String(item.auth || "sql"),
    String(item.username || "").trim().toLowerCase(),
    String(item.database || "master").trim().toLowerCase(),
  ].join("|");
}

export function publicProfile(item: Profile) {
  const auth = String(item.auth || "sql");
  let label = `${item.server || "server"} · ${item.database || "master"}`;
  if (item.instance) label = `${item.server}\\${item.instance} · ${item.database || "master"}`;
  if (auth === "windows") label += " · Windows";
  else if (item.username) label += ` · ${item.username}`;
  return {
    id: item.id,
    label,
    server: item.server || "",
    port: item.port || 1433,
    instance: item.instance || "",
    auth,
    username: item.username || "",
    database: item.database || "master",
    encrypt: Boolean(item.encrypt),
    has_password: Boolean(item.password),
    remember_password: Boolean(item.remember_password),
    last_used: item.last_used || "",
  };
}

export function listProfiles() {
  return load()
    .sort((a, b) => String(b.last_used || "").localeCompare(String(a.last_used || "")))
    .map(publicProfile);
}

export function getProfile(id: string) {
  return load().find((item) => item.id === id) || null;
}

export function readPassword(item: Profile) {
  return decryptSecret(String(item.password || ""));
}

export function upsertProfile(payload: Profile, rememberPassword = false) {
  const items = load();
  const incoming: Profile = {
    server: String(payload.server || "").trim(),
    port: Number(payload.port || 1433),
    instance: String(payload.instance || "").trim(),
    auth: payload.auth === "windows" ? "windows" : "sql",
    username: String(payload.username || "").trim(),
    database: String(payload.database || "master").trim() || "master",
    encrypt: Boolean(payload.encrypt),
    remember_password: rememberPassword,
    last_used: new Date().toISOString().replace("T", " ").slice(0, 19),
  };
  const password = String(payload.password || "");
  let match = items.find((item) => keyOf(item) === keyOf(incoming));
  if (!match) {
    incoming.id = randomBytes(6).toString("hex");
    if (rememberPassword && password) incoming.password = encryptSecret(password);
    items.unshift(incoming);
    match = incoming;
  } else {
    Object.assign(match, incoming);
    if (rememberPassword) {
      if (password) match.password = encryptSecret(password);
    } else {
      delete match.password;
      match.remember_password = false;
    }
  }
  save(items);
  return publicProfile(match);
}

export function deleteProfile(id: string) {
  const items = load();
  const kept = items.filter((item) => item.id !== id);
  if (kept.length === items.length) return false;
  save(kept);
  return true;
}
